### PR TITLE
[KVCache] Add max num threads awareness to KVCache kernels

### PR DESCRIPTION
**Pull request by CharlieFRuan**
WebGPU has limit on `bdx * bdy * bdz <= 256` and `bdz <= 256`. Prior to this PR, many kernels would break the limit, depending on the model's group size, head_dim, etc.

This PR adds a helper method `check_max_num_threads()` that checks for such error during compilation, otherwise we only find out about this issue when actually deploying in the browser. We also add TinyLlama to presets, hence the CI.

We also backoff `bdx`, `bdy`, `bdz` based on each kernel's parallelization strategy if needed.

Co-authored-by: Ruihang Lai <ruihangl@cs.cmu.edu>

### CODE REVIEW

File: python/mlc_chat/nn/kv_cache.py [hunk 1]
```diff
@@ -477,10 +479,20 @@ def _attention_prefill(h_kv, h_q, d, dtype, target: Target):  # pylint: disable=
     group_size = h_q // h_kv
     sm_scale = 1.0 / math.sqrt(float(d)) * math.log2(math.exp(1))
 
+    bdx = 32
```
Comment (CharlieFRuan) [inline python/mlc_chat/nn/kv_cache.py:482]: I'm not exactly sure about substituting `32` with `bdx` in the entire `_attention_prefill()` and `_attention_prefill_ragged()`; is it the intended semantics?

File: python/mlc_chat/nn/kv_cache.py [hunk 2]
```diff
@@ -1049,6 +1062,9 @@ def _merge_state_inplace(
     VEC_SIZE = min(max(8 // v_dtype_bytes, head_dim // 32), 4)
     bdx = head_dim // VEC_SIZE
     bdy = num_heads
+    while bdx * bdy > target.max_num_threads and bdy > 1:
+        bdy //= 2
+    check_max_num_threads(target, bdx=bdx, bdy=bdy, bdz=1)
```
Comment (CharlieFRuan) [inline python/mlc_chat/nn/kv_cache.py:1067]: Will this affect accuracy?

**Review by CharlieFRuan — COMMENTED** [review]
The kernels changed are `_merge_state_inplace()`, `_attention_prefill()`, `_attention_prefill_ragged()`, and `tir_rotary()`.

Though I tested for accuracy on WebGPU for q4f16_1 llama7b, 13b, gema 2b, tinyLlama, and all had no issue.

File: python/mlc_chat/support/max_thread_check.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+from tvm.target import Target
+
+
+def check_max_num_threads(target: Target, bdx: int, bdy: int, bdz: int):
+    """Check whether max num threads exceeded given a target."""
+    assert (
+        bdx * bdy * bdz <= target.max_num_threads
+    ), f"{target.kind} max num threads exceeded: {bdx}*{bdy}*{bdz}>{target.max_num_threads}"
+
+    if str(target.kind) != "webgpu":
```
Comment (MasterJH5574) [inline python/mlc_chat/support/max_thread_check.py:12]: Should it be `== "webgpu"`? Also for WebGPU `blockIdx.z` should be 1, which is also be something we could check.

File: python/mlc_chat/nn/kv_cache.py (hunk 2 above)
Comment (MasterJH5574) [reply python/mlc_chat/nn/kv_cache.py:1067]: Simply halving `bdy` might not be correct. May need to change the TIR as well.

File: python/mlc_chat/support/max_thread_check.py (hunk 3 above)
Comment (CharlieFRuan) [reply python/mlc_chat/support/max_thread_check.py]: Resolved, thank you!

File: python/mlc_chat/nn/kv_cache.py (hunk 2 above)
Comment (CharlieFRuan) [reply python/mlc_chat/nn/kv_cache.py]: Updated with a gdy accordingly

**Pull request merged**